packages/flutter/lib/src/widgets/scrollable.dart

```diff
@@ -1605,3 +1613,668 @@ class _RestorableScrollOffset extends RestorableValue<double?> {
+/// To manipulate one of the internal [Scrollable] widget's scroll position, use
+/// the object obtained from the [verticalScrollable] or [horizontalScrollable]
+/// property.
+///
+/// To be informed of when a [TwoDimensionalScrollable] widget is scrolling,
+/// use a [NotificationListener] to listen for [ScrollNotification]s.
+/// Both axes will have the same viewport depth since there is only one
+/// viewport, and so should be differentiated by the [Axis] of the
+/// [ScrollMetrics] provided by the notification.
+class TwoDimensionalScrollableState extends State<TwoDimensionalScrollable> {
+  late ScrollBehavior _configuration;
+  ScrollController? _verticalFallbackController;
+  ScrollController? _horizontalFallbackController;
+  final GlobalKey<ScrollableState> _verticalOuterScrollableKey = GlobalKey<ScrollableState>();
+  final GlobalKey<ScrollableState> _horizontalInnerScrollableKey = GlobalKey<ScrollableState>();
+
+  /// The [ScrollableState] of the vertical axis.
+  ///
+  /// Accessible by calling [TwoDimensionalScrollable.of].
+  ///
+  /// Alternatively, [Scrollable.of] can be used by providing [Axis.vertical]
+  /// to the `axis` parameter.
+  ScrollableState get verticalScrollable {
+    assert(_verticalOuterScrollableKey.currentState != null);
+    return _verticalOuterScrollableKey.currentState!;
+  }
+
+  /// The [ScrollableState] of the horizontal axis.
+  ///
+  /// Accessible by calling [TwoDimensionalScrollable.of].
+  ///
+  /// Alternatively, [Scrollable.of] can be used by providing [Axis.horizontal]
+  /// to the `axis` parameter.
+  ScrollableState get horizontalScrollable {
+    assert(_horizontalInnerScrollableKey.currentState != null);
+    return _horizontalInnerScrollableKey.currentState!;
+  }
+
+  @override
+  void initState() {
+    if (widget.verticalDetails.controller == null) {
+      _verticalFallbackController = ScrollController();
+    }
+    if (widget.horizontalDetails.controller == null) {
+      _horizontalFallbackController = ScrollController();
+    }
+    super.initState();
+  }
+
+  @override
+  void didChangeDependencies() {
+    _configuration = ScrollConfiguration.of(context);
+    super.didChangeDependencies();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    assert(axisDirectionToAxis(widget.horizontalDetails.direction) == Axis.horizontal);
+    assert(axisDirectionToAxis(widget.verticalDetails.direction) == Axis.vertical);
+
+    Widget result = _VerticalOuterDimension(
+      key: _verticalOuterScrollableKey,
+      axisDirection: widget.verticalDetails.direction,
+      controller: widget.verticalDetails.controller ?? _verticalFallbackController!,
+      physics: widget.verticalDetails.physics,
+      clipBehavior: widget.verticalDetails.clipBehavior ?? Clip.hardEdge,
+      incrementCalculator: widget.incrementCalculator,
+      excludeFromSemantics: widget.excludeFromSemantics,
+      restorationId: widget.restorationId == null ? null : 'OuterTwoDimensionalScrollable',
```

Comment:
here and below: You can unconditionally wrap everything in a RestorationScope and then unconditionally pass in OuterTwoDimensionalScrollable and InnerTwoDimensionalScrollable below.